### PR TITLE
mbc iec refactor

### PR DESCRIPTION
**Pull request by jd-lara**
This refactore relies on InfrastructureSystems#565

### CODE REVIEW

File: src/models/cost_functions/ImportExportTimeSeriesCost.jl [hunk 1]
```diff
@@ -0,0 +1,68 @@
+"""
+$(TYPEDEF)
```
Comment (luke-kiernan) [inline src/models/cost_functions/ImportExportTimeSeriesCost.jl:2]: I haven't noticed these `TYPEDEF` things before. I don't see a definition of that symbol anywhere...docstring formatting thing?

File: src/models/cost_functions/ImportExportCost.jl [hunk 2]
```diff
@@ -136,35 +124,29 @@ function make_import_export_curve(
...
-
-# Examples
-```julia
-ic = make_import_curve([0.0, 100.0, 105.0, 120.0, 130.0], [25.0, 26.0, 28.0, 30.0])
-ic1 = make_import_curve(power = [0.0, 100.0, 105.0, 120.0, 130.0], price = [25.0, 26.0, 28.0, 30.0], power_units = UnitSystem.NATURAL_UNITS)
-```
+Make a time-series-backed `CostCurve{TimeSeriesPiecewiseIncrementalCurve}` from a
+`TimeSeriesKey`.
 """
+function make_import_export_curve(
```
Comment (luke-kiernan) [inline src/models/cost_functions/ImportExportCost.jl]: Confusing: I don't see anything import-export specific in the body, just `CostCurve` and `TimeSeriesPiecewiseIncrementalCurve`. Also, I thought we'd switched to a different struct when the curves are time-varying.

File: src/models/cost_functions/ImportExportCost.jl [hunk 3]
```diff
@@ -85,8 +77,8 @@ function is_import_export_curve(curve::ProductionVariableCostCurve)
 end
 
 """
-Make a CostCurve{PiecewiseIncrementalCurve} suitable for inclusion in a ImportExportCost from
-the FunctionData that might be used to store such a cost curve in a time series.
+Make a static `CostCurve{PiecewiseIncrementalCurve}` suitable for an `ImportExportCost`
+from `PiecewiseStepData`.
 """
 function make_import_export_curve(
```
Comment (luke-kiernan) [inline src/models/cost_functions/ImportExportCost.jl:80]: Why so many constructors? Taking census:
- 5 variations on `make_import_export_curve`: function data, powers/prices (2 variations: vector vs floats), all kwargs, and time series key
- 3 variations on `make_import_curve`: powers/prices (2 variations), all kwargs
- 3 variations on `make_export_curve`: powers/prices (2 variations), all kwargs

Do we really need all of these?

File: src/models/cost_functions/MarketBidCost.jl [hunk 4]
```diff
@@ -243,26 +142,29 @@ function make_market_bid_curve(powers::Vector{Float64},
...
+    input_at_zero::Union{Nothing, Float64} = nothing,
+)
     cc = CostCurve(IncrementalCurve(data, initial_input, input_at_zero), power_units)
     @assert is_market_bid_curve(cc)
     return cc
 end
 
 """
-Auxiliary make market bid curve for timeseries with nothing inputs.
+Make a time-series-backed `CostCurve{TimeSeriesPiecewiseIncrementalCurve}` from
```
Comment (luke-kiernan) [inline src/models/cost_functions/MarketBidCost.jl]: Same comment as above: is this really MBC specific? And if the MBC definition in this file is only for static data, why do we have time-dependent stuff here?

File: test/test_cost_functions.jl [hunk 5]
```diff
@@ -111,365 +111,160 @@ test_costs = Dict(
...
-    data_float = SortedDict(initial_time => test_costs[Float64])
-    forecast_fd = IS.Deterministic("fuel_cost", data_float, resolution)
-    set_fuel_cost!(sys, generator, forecast_fd)
-    fuel_forecast = get_fuel_cost(generator; start_time = initial_time)
-    @test first(TimeSeries.values(fuel_forecast)) == first(data_float[initial_time])
-    fuel_forecast = get_fuel_cost(generator)  # missing start_time filled in with initial time
-    @test first(TimeSeries.values(fuel_forecast)) == first(data_float[initial_time])
-end
-@testset "Test MarketBidCost no-load cost (single number and time series)" begin
+@testset "Test MarketBidCost no_load_cost and shut_down are LinearCurve" begin
```
Comment (luke-kiernan) [inline test/test_cost_functions.jl]: AI test spam: sure have *some* test run the kwarg constructor, but this seems like overkill.

File: test/test_cost_functions.jl [hunk 6]
```diff
@@ -111,365 +111,160 @@ test_costs = Dict(
...
-
-    set_no_load_cost!(sys, generator, forecast_fd)
-    @test first(TimeSeries.values(get_no_load_cost(generator, op_cost))) ==
-          first(data_float[initial_time])
+    set_shut_down!(op_cost, LinearCurve(7.0))
+    @test get_shut_down(op_cost) == LinearCurve(7.0)
 end
 
-@testset "Test MarketBidCost startup cost (single number, tuple, and time series)" begin
+@testset "Test MarketBidCost start_up setters" begin
```
Comment (luke-kiernan) [inline test/test_cost_functions.jl]: more AI test spam.

File: test/test_cost_functions.jl [hunk 7]
```diff
@@ -111,365 +111,160 @@ test_costs = Dict(
...
-    service_data = data_pwl
-    sys = PSB.build_system(PSITestSystems, "test_RTS_GMLC_sys")
-    generator = get_component(ThermalStandard, sys, "322_CT_6")
-    market_bid = MarketBidCost(nothing)
-    set_operation_cost!(generator, market_bid)
-    forecast_fd = IS.Deterministic(
-        "decremental_variable_cost_function_data",
-        Dict(k => get_function_data.(v) for (k, v) in pairs(data_pwl)),
-        resolution,
+    # Test setting incremental variable cost
```
Comment (luke-kiernan) [inline test/test_cost_functions.jl]: Again I question if this is necessary, borderline AI test spam.

File: test/test_cost_functions.jl [hunk 8]
```diff
@@ -111,365 +111,160 @@ test_costs = Dict(
...
-    iocs = get_decremental_offer_curves(generator, market_bid)
-    isequal(first(TimeSeries.values(iocs)), first(data_pwl[initial_time]))
-    cost_forecast =
-        get_decremental_variable_cost(generator, market_bid; start_time = initial_time)
-    @test isequal(first(TimeSeries.values(cost_forecast)), first(data_pwl[initial_time]))
+    @test_throws ArgumentError set_variable_cost!(
+        sys, generator, cc_system, UnitSystem.NATURAL_UNITS)
 end
 
 @testset "Test `MarketBidCost` with single `start_up` value" begin
```
Comment (luke-kiernan) [inline test/test_cost_functions.jl]: AI test spam

File: test/test_cost_functions.jl [hunk 9]
```diff
@@ -111,365 +111,160 @@ test_costs = Dict(
...
 
-@testset "Test MarketBidCost shutdown cost (single number and time series)" begin
+@testset "Test ReserveDemandCurve nothing constructor" begin
+    reserve = ReserveDemandCurve{ReserveUp}(nothing)
+    @test get_name(reserve) == "init"
+    @test get_available(reserve) == false
+    @test get_variable(reserve) == PSY.ZERO_OFFER_CURVE
+end
+
+@testset "Test fuel cost (scalar and time series)" begin
```
Comment (luke-kiernan) [inline test/test_cost_functions.jl]: borderline AI test spam

**Review by luke-kiernan — CHANGES_REQUESTED** [review]
Broad ideas look fine. Implementation looks like unrefined AI code. 2 big picture things:
- eliminate dumb test cases
- prune some of the constructors: we don't need one to cover every possible scenario arument-wise.

**Comment by luke-kiernan** [comment]
> - Move TS-backed curve constructors out of static cost files
>   (`make_market_bid_ts_curve` to MarketBidTimeSeriesCost.jl,
>   `make_import_export_ts_curve` to ImportExportTimeSeriesCost.jl)
> - Prune `make_import*_curve` constructors from 11 methods to 5 positional
>   forms; drop the unused public `make_import_export_curve` API surface
> - Trim AI-generated test spam in test_cost_functions.jl
> - Add resolution tests for `MarketBidTimeSeriesCost` and
>   `ImportExportTimeSeriesCost` covering get_*_variable_cost at start_time

I decided to go to positional args for type stability; I'm open to adding more constructors, as the need arises. I added tests for the substantial pieces of CoPilot's comments, though I didn't cover the `start = nothing` edge cases.

**Review by luke-kiernan — APPROVED** [review]

File: test/test_cost_functions.jl [hunk 10]
```diff
@@ -491,10 +459,31 @@ end
...
-# TODO: ReserveDemandTimeSeriesCurve resolution test deferred — constructing the
-# reserve requires a TimeSeriesKey, which today can only be obtained via
-# `add_time_series!` on a component already in a system. Pending an IS interface
-# for constructing a TimeSeriesKey upfront (or a `set_variable_from_forecast!`
-# helper), the test would have to do an ugly bootstrap-and-swap dance.
+@testset "ReserveDemandTimeSeriesCurve resolves variable cost at start_time" begin
+    sys = PSB.build_system(PSITestSystems, "test_RTS_GMLC_sys")
+    # `ForecastKey` identifies a time series by name/type/timing — it does not
+    # bind to a specific component, so we can bootstrap a key from a throwaway
+    # generator and reuse it after attaching the same forecast to the reserve.
```
Comment (luke-kiernan) [inline test/test_cost_functions.jl:471]: Ick. I guess this is fine for now, but we should really come up with a cleaner way.

**Review by luke-kiernan — APPROVED** [review]
